File: rootfs/fullapp/runtimes/python3/workload.py
#!/usr/bin/env python3

from importlib import import_module
import os
import struct
import json
from subprocess import run, Popen
import time
import socket

# vsock to communicate with the host
VSOCKPORT = 1234
sock = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
hostaddr = (socket.VMADDR_CID_HOST, VSOCKPORT)

app = import_module('workload')

# for function snapshot
for i in range(1, os.cpu_count()):
    Popen('taskset -c %d outl 124 0x3f0'%(i), shell=True)
run('taskset -c 0 outl 124 0x3f0', shell=True)

sock.connect(hostaddr)
while True:
    data = sock.recv(4, socket.MSG_WAITALL)
    res = struct.unpack(">I", data)
    requestJson = sock.recv(res[0], socket.MSG_WAITALL).decode("utf-8")

    request = json.loads(requestJson)

    start = time.monotonic_ns()
    response = app.handle(request)
    response['duration'] = time.monotonic_ns() - start

    responseJson = json.dumps(response)

    sock.sendall(struct.pack(">I", len(responseJson)))
    sock.sendall(bytes(responseJson, "utf-8"))

File: rootfs/fullapp/runtimes/nodejs/workload.js
const vsock = require("vsock");
const { execSync, exec } = require("child_process");

module.paths.push("/srv/node_modules");
const app = require("/srv/workload");

const cpu_count = require("os").cpus().length;
for (var i = 1; i < cpu_count; i++) {
    exec(`taskset -c ${i} do-snapshot 123`);
}
execSync('taskset -c 0 do-snapshot 123')

const sock_conn = vsock.connect(2, 1234);

(async function() {
  while (true) {
    const req = await vsock.readRequest(sock_conn);

    const hrstart = process.hrtime();
    const resp = await app.handle(req);
    const hrend = process.hrtime(hrstart)
    resp.runtime_sec = hrend[0];
    resp.runtime_ms = hrend[1] / 1000000;
    await vsock.writeResponse(sock_conn, resp);
  }
})();

